File: packages/python/src/typeflux/contracts/cache.py
"""Cross-run cache key + record contract (#391).

Defines the key a cross-run ``CacheStore`` (E3, #398), adopter cache implementations, and
the TypeScript SDK all use to memoize an AI-activity result across runs. The
provider-side *prefix* cache (``session_cache_identity``, #60) is a separate
concern and is not involved here.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from dataclasses import dataclass, field
from hashlib import sha256
from typing import Any

from typeflux.manifests._common import canonical_json

__all__ = [
    "CacheKey",
    "cache_input_hash",
    "cache_key_digest",
    "cache_record",
]


@dataclass(frozen=True)
class CacheKey:
    """Identifies a cached AI-activity result.

    ``activity`` is the activity ("node") name; ``input_hash`` is the digest of
    the call-determining inputs (see :func:`cache_input_hash`); ``scope`` is a
    generic string map for tenancy/partitioning (e.g.
    ``{"company_id": ..., "product_id": ...}`` — empty means global). Keeping
    ``scope`` generic lets any deployment choose its partition keys while
    supporting the ``(node, input_hash, company_id, product_id)`` shape used by adopter caches.
    """

    activity: str
    input_hash: str
    scope: Mapping[str, str] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "activity": self.activity,
            "input_hash": self.input_hash,
            "scope": dict(self.scope),
        }

    def digest(self) -> str:
        """A single stable key string for stores that want a flat key."""
        return cache_key_digest(self)


def cache_key_digest(key: CacheKey) -> str:
    """``sha256`` over the key's canonical JSON. ``canonical_json`` sorts keys,
    so the digest is independent of ``scope`` insertion order."""

    return sha256(canonical_json(key.to_dict()).encode("utf-8")).hexdigest()


def cache_input_hash(
    *,
    activity: str,
    input_schema_hash: str,
    rendered_messages_hash: str,
    provider_params: Mapping[str, Any] | None = None,
    artifacts: Sequence[Mapping[str, Any]] | None = None,
) -> str:
    """Digest of the inputs that determine an AI-activity call.

    Reuses the frozen execution-manifest hashes (#390): ``rendered_messages_hash``
    encodes the prompt rendered with the input, so identical input + prompt +
    behavior params yield the same hash — one source of truth for "what
    determines a call". Mirrors an adopter cache's input hash (content hash of files
    + prompt + input).

    ``artifacts`` is the resolved-artifact identity (#504) in the
    ``artifact_groups_cache_identity`` shape (per artifact: group/index/
    source_kind/kind/media_type/role/sha256/size_bytes with ``None`` dropped,
    plus the SOURCE — url/uri/provider+file_id — whenever no sha256 pins the
    bytes; empty groups excluded). An artifact part RENDERS as only its group
    name + preamble text, so the rendered-messages hash cannot see the
    underlying bytes — without this fold, replacing a file (or the source
    location of an unhashed artifact) under the same group name would serve a
    stale cached output. Omitted from the payload when absent or empty, so keys
    for artifact-free calls (and all previously stored entries) are unchanged.
    """

    payload: dict[str, Any] = {
        "activity": activity,
        "input_schema_hash": input_schema_hash,
        "rendered_messages_hash": rendered_messages_hash,
        "provider_params": dict(provider_params) if provider_params else {},
    }
    if artifacts:
        payload["artifacts"] = [dict(group) for group in artifacts]
    return sha256(canonical_json(payload).encode("utf-8")).hexdigest()


def cache_record(
    *,
    key: CacheKey,
    output: Any,
    created_at: str,
    output_schema_hash: str,
    manifest_hash: str | None = None,
    tokens_saved: int | None = None,
    subjects: Sequence[str] | None = None,
) -> dict[str, Any]:
    """A stored cache entry: the cached validated output plus provenance.

    ``output`` is the activity's validated output (JSON-serializable);
    ``created_at`` is an ISO-8601 timestamp; ``output_schema_hash`` ties the
    record to the output schema it was produced against (so a schema change can
    invalidate stale entries). ``manifest_hash`` / ``tokens_saved`` are optional
    provenance.

    ``subjects`` (#715 slice 1) records the subject id(s) whose data produced
    this entry, so a later per-subject cache invalidation can find it. It lives on
    the RECORD, NOT in ``CacheKey`` — folding it into the key would change the
    digest and over-partition scope-keyed (``{company_id, product_id}``) entries,
    wrecking the hit rate. Present-only: absent when empty, so records for
    subject-free calls (and every previously stored entry) are byte-unchanged and
    the key digest is untouched.
    """

    record: dict[str, Any] = {
        "key": key.to_dict(),
        "output": output,
        "created_at": created_at,
        "output_schema_hash": output_schema_hash,
    }
    if manifest_hash is not None:
        record["manifest_hash"] = manifest_hash
    if tokens_saved is not None:
        record["tokens_saved"] = tokens_saved
    if subjects:
        record["subjects"] = list(subjects)
    return record

File: packages/typescript/temporal/src/cache.ts
/**
 * Cross-run cache key + record contract (#391), mirroring the Python
 * `contracts/cache.py`. The key a cross-run cache store, adopter cache implementations, and
 * this SDK all use to memoize an AI-activity result across runs.
 */

import { createHash } from "node:crypto";

import { canonicalJson } from "./canonical-json.js";

/** Identifies a cached AI-activity result. */
export interface CacheKey {
  /** The activity ("node") name. */
  activity: string;
  /** Digest of the call-determining inputs (see {@link cacheInputHash}). */
  input_hash: string;
  /** Tenancy/partition keys (e.g. `{ company_id, product_id }`); empty = global. */
  scope: Record<string, string>;
}

/** A stored cache entry: the cached validated output plus provenance. */
export interface CacheRecord {
  key: CacheKey;
  output: unknown;
  /** ISO-8601 timestamp. */
  created_at: string;
  /** Ties the entry to the output schema it was produced against. */
  output_schema_hash: string;
  manifest_hash?: string;
  tokens_saved?: number;
  /**
   * Subject id(s) whose data produced this entry (#715 slice 1) — so a later
   * per-subject cache invalidation can find it. Lives on the RECORD, NOT in the
   * {@link CacheKey}: folding it into the key would change the digest and
   * over-partition scope-keyed entries. Present-only (absent when empty), so
   * records for subject-free calls and every previously stored entry are
   * byte-unchanged and the key digest is untouched.
   */
  subjects?: string[];
}

function sha256Hex(value: string): string {
  return createHash("sha256").update(value, "utf-8").digest("hex");
}

/** Construct a {@link CacheKey} (scope defaults to global). */
export function cacheKey(
  activity: string,
  inputHash: string,
  scope: Record<string, string> = {},
): CacheKey {
  return { activity, input_hash: inputHash, scope: { ...scope } };
}

/**
 * `sha256` over the key's canonical JSON. `canonicalJson` sorts keys, so the
 * digest is independent of `scope` insertion order.
 */
export function cacheKeyDigest(key: CacheKey): string {
  return sha256Hex(
    canonicalJson({
      activity: key.activity,
      input_hash: key.input_hash,
      scope: { ...key.scope },
    }),
  );
}

/**
 * Digest of the inputs that determine an AI-activity call. `renderedMessagesHash`
 * (frozen by the execution-manifest contract, #390) encodes the prompt rendered
 * with the input, so identical input + prompt + behavior params yield the same
 * hash. Mirrors an adopter cache's input hash.
 *
 * `artifacts` is the resolved-artifact identity (#504) in the
 * `artifactGroupsCacheIdentity` shape (per artifact: group/index/source_kind/
 * kind/media_type/role/sha256/size_bytes with absent fields dropped, plus the
 * SOURCE — url/uri/provider+file_id — whenever no sha256 pins the bytes; empty
 * groups excluded). An artifact part RENDERS as only its group name + preamble
 * text, so the rendered-messages hash cannot see the underlying bytes — without
 * this fold, replacing a file (or the source location of an unhashed artifact)
 * under the same group name would serve a stale cached output. Omitted from the
 * payload when absent or empty (Python `if artifacts:` parity), so keys for
 * artifact-free calls (and all previously stored entries) are unchanged.
 */
export function cacheInputHash(params: {
  activity: string;
  inputSchemaHash: string;
  renderedMessagesHash: string;
  providerParams?: Record<string, unknown> | null;
  artifacts?: readonly Record<string, unknown>[] | null;
}): string {
  const payload: Record<string, unknown> = {
    activity: params.activity,
    input_schema_hash: params.inputSchemaHash,
    rendered_messages_hash: params.renderedMessagesHash,
    provider_params: params.providerParams ? { ...params.providerParams } : {},
  };
  if (params.artifacts != null && params.artifacts.length > 0) {
    payload["artifacts"] = params.artifacts.map((group) => ({ ...group }));
  }
  return sha256Hex(canonicalJson(payload));
}

/**
 * Build a {@link CacheRecord}. `manifestHash` / `tokensSaved` are optional
 * provenance and are omitted from the record when not provided (matching the
 * Python `cache_record`).
 */
export function cacheRecord(params: {
  key: CacheKey;
  output: unknown;
  createdAt: string;
  outputSchemaHash: string;
  manifestHash?: string | null;
  tokensSaved?: number | null;
  subjects?: readonly string[] | null;
}): CacheRecord {
  const record: CacheRecord = {
    key: {
      activity: params.key.activity,
      input_hash: params.key.input_hash,
      scope: { ...params.key.scope },
    },
    output: params.output,
    created_at: params.createdAt,
    output_schema_hash: params.outputSchemaHash,
  };
  if (params.manifestHash != null) {
    record.manifest_hash = params.manifestHash;
  }
  if (params.tokensSaved != null) {
    record.tokens_saved = params.tokensSaved;
  }
  if (params.subjects != null && params.subjects.length > 0) {
    record.subjects = [...params.subjects];
  }
  return record;
}
